File: 1717-maximum-score-from-removing-substrings/1717-maximum-score-from-removing-substrings.cpp
#include <string.h>

class Solution {
public:
    int maximumGain(string s, int x, int y) {
        int ans = 0;
        string high_prior, low_prior;
        int score[2];

        if (x > y)
        {
            high_prior = "ab";
            low_prior = "ba";
            score[0] = x;
            score[1] = y;
        }
        else
        {
            high_prior = "ba";
            low_prior = "ab";
            score[0] = y;
            score[1] = x;
        }

        deque<char> first;
        for (int i=0; i<s.length(); i++)
        {
            if (first.empty())
            {
                first.push_back(s[i]);
            }
            else
            {
                if (first.back()==high_prior[0] && s[i]==high_prior[1])
                {
                    first.pop_back();
                }
                else
                {
                    first.push_back(s[i]);
                }
            }
        }

        ans = (s.length()-first.size())*score[0]/2;

        stack<char> second;
        int first_size = first.size();
        for (int i=0; i<first_size; i++)
        {
            char tmp = first.front();
            first.pop_front();
            if (second.empty())
            {
                second.push(tmp);
            }
            else
            {
                if (second.top()==low_prior[0] && tmp==low_prior[1])
                {
                    second.pop();
                }
                else
                {
                    second.push(tmp);
                }
            }
        }

        ans += (first_size-second.size())*score[1]/2;
        return ans;
    }
};

File: 1717-maximum-score-from-removing-substrings/1717-maximum-score-from-removing-substrings.py
class Solution:
    def maximumGain(self, s: str, x: int, y: int) -> int:
        ans = 0
        stack = deque()

        if x>y:
            high_prior = "ab"
            low_prior = "ba"
            score = [x, y]
        else:
            high_prior = "ba"
            low_prior = "ab"
            score = [y, x]

        for x in s:
            if len(stack)==0:
                stack.append(x)
            else:
                if stack[-1]+x==high_prior:
                    stack.pop()
                else:
                    stack.append(x)
        ans = ((len(s)-len(stack))//2)*score[0]

        stack2 = deque()
        for x in stack:
            if len(stack2)==0:
                stack2.append(x)
            else:
                if stack2[-1]+x==low_prior:
                    stack2.pop()
                else:
                    stack2.append(x)
        ans += ((len(stack)-len(stack2))//2)*score[1]

        return ans